### ISSUE TITLE
Google Drive: Files in explorer are wrongly assumed to be directories

### ISSUE THREAD

**Issue by LeNouveau** [microsoft/vscode#69333]
When opening a `*.gsheet` file synchronized with Google Drive, VSCode loops and is partially blocked. I can't use it anymore. 
A simple reboot solves the looping. 

Other behaviour : if a file is already opened, on opening the `*.gsheet` file, I have a message 'File is a Folder', but no looping. 

For reminder, when clicking on this kind of file, Google Drive redirects us on the web interface. 
The file isn't really accessible in local. 

[Demo of behaviour](https://transfer.sh/10q9RF/looping.mp4)

Versions : 

- Version: 1.32.0-insider (system setup)
- Commit: aeaef41d51201e555735f5e8d2f38a9d0ddb9026
- OS: Windows_NT x64 10.0.17763

Steps to Reproduce:

1. Install "Google Drive File Stream". A `g:\` device should appear. 
2. Through Google Drive web interface, create a `*.gsheet` file. 
3. Open `g:\` folder with VSCode. 
4. Click on the `*.gsheet` file. 

Does this issue occur when all extensions are disabled ? : Yes

**Comment by bpasero** [comment]
I cannot open the video. What does looping mean?

**Comment by LeNouveau** [comment]
When I click on the file, inside the tab "Open Editor", the file appears, disappears, appears, disappears, etc... 
I can't select another file any more. 

I think it's linked to the g:\ partition, created by Google Drive File Stream. 
Another weird behaviour : when creating a file from VSCode, when asking for the filename, its disappears. The area is greyed, but there is no input field any more.

**Comment by bpasero** [comment]
Please see if you can find any logged output that could help us find the source of the issue:
* run `code --verbose` from the command line and attach the output
* in Code select **Help | Toggle Developer Tools** and select the **Console**
* attach any output you see in the **Console** you can find

Thanks!

**Comment by LeNouveau** [comment]
Sure, 
Logs : https://gist.github.com/LeNouveau/9dcc19102861395bc28ce9dd3166adfb

**Comment by bpasero** [comment]
Can you try to reproduce with our nightly insider builds? You can give our preview releases a try from: https://code.visualstudio.com/insiders/

**Comment by bpasero** [comment]
@LeNouveau thanks I can see what is going on but I do not have an understanding why this happens:
* you click on a file in the file explorer in VSCode
* we try to open it as editor
* we try to resolve the file contents to show the editor
* somehow we get an error from node.js that the file is actually a folder
* we close the editor and instead try to focus the folder in the explorer
* this triggers an opening again of the file from the explorer
* we end up in an endless loop

@isidorn would it be possible to get a new method from explorer service to either:
* select but not open on select
* find out the current selected resource (and then I can add the logic to not call this over and over again)

@LeNouveau could you try to use node.js to stat the path and see if the stat also tells it is a folder?

**Comment by isidorn** [comment]
@isidorn if you just call `explorerService.select` that should select it and not open it.

**Comment by bpasero** [comment]
@isidorn that is not what I see, I do this:

https://github.com/Microsoft/vscode/blob/804373ac4d5c84e96c8086cee5b9be132df88837/src/vs/workbench/contrib/files/browser/editors/textFileEditor.ts#L238

**Steps:**
* return `FileOperationResult.FILE_IS_DIRECTORY` from `IFileService#resolveStreamContent`
* open a file from the explorer
* get an endless loop

**Comment by LeNouveau** [comment]
Already on Insider.
Version: 1.32.0-insider (system setup)
Commit: 804373ac4d5c84e96c8086cee5b9be132df88837
Date: 2019-03-01T18:22:35.336Z
Electron: 3.1.3
Chrome: 66.0.3359.181
Node.js: 10.2.0
V8: 6.6.346.32
OS: Windows_NT x64 10.0.17763

Still have the problem.

Le ven. 1 mars 2019 à 15:47, Benjamin Pasero <notifications@github.com> a
écrit :

**Comment by bpasero** [comment]
@LeNouveau could you try to use node.js to stat the path and see if the stat also tells it is a folder?

**Comment by LeNouveau** [comment]
Is it what you expected ?

> fs.statSync("Simple gsheet file.gsheet")
Stats {
  dev: 428019990,
  mode: 33206,
  nlink: 0,
  uid: 0,
  gid: 0,
  rdev: 0,
  blksize: undefined,
  ino: 0,
  size: 1,
  blocks: undefined,
  atimeMs: 1551282892782,
  mtimeMs: 1551282892782,
  ctimeMs: 1551282892782,
  birthtimeMs: 1551282856784,
  atime: 2019-02-27T15:54:52.782Z,
  mtime: 2019-02-27T15:54:52.782Z,
  ctime: 2019-02-27T15:54:52.782Z,
  birthtime: 2019-02-27T15:54:16.784Z }


Le sam. 2 mars 2019 à 11:17, Benjamin Pasero <notifications@github.com> a
écrit :

**Comment by bpasero** [comment]
@LeNouveau yeah, can you also do `fs.statSync("Simple gsheet file.gsheet").isDirectory()`?

**Comment by LeNouveau** [comment]
```
> fs.statSync("Simple gsheet file.gsheet").isDirectory()
false
```

Le sam. 2 mars 2019 à 14:51, Benjamin Pasero <notifications@github.com> a
écrit :

**Comment by bpasero** [comment]
@LeNouveau which node version is that? Can you try with `10.2.0`? That is the one VSCode currently uses.

**Comment by isidorn** [comment]
@bpasero I just tried to repro and calling 			this.explorerService.select(uri, true);
does not open the uri for me in the editor.
Can you please file a new issue for me if you think that something is wrong on my side? Thanks a lot!

**Comment by bpasero** [comment]
@isidorn do you get the endless loop from https://github.com/Microsoft/vscode/issues/69333#issuecomment-468745958

**Comment by LeNouveau** [comment]
@bpasero 

```
PS G:\Mon Drive\test> node --version
v10.2.0
PS G:\Mon Drive\test> node
> fs.statSync("Simple gsheet file.gsheet")
Stats {
  dev: 428019990,
  mode: 33206,
  nlink: 0,
  uid: 0,
  gid: 0,
  rdev: 0,
  blksize: undefined,
  ino: 0,
  size: 1,
  blocks: undefined,
  atimeMs: 1551282892782,
  mtimeMs: 1551282892782,
  ctimeMs: 1551282892782,
  birthtimeMs: 1551282856784,
  atime: 2019-02-27T15:54:52.782Z,
  mtime: 2019-02-27T15:54:52.782Z,
  ctime: 2019-02-27T15:54:52.782Z,
  birthtime: 2019-02-27T15:54:16.784Z }
> fs.statSync("Simple gsheet file.gsheet").isDirectory()
false
```

**Comment by isidorn** [comment]
@bpasero yes I can.
The issue is that [here](https://github.com/Microsoft/vscode/blob/0a1e5ea6d12272f164784b15a2d9d1eaabfed609/src/vs/workbench/contrib/files/browser/editors/textFileEditor.ts#L237) you are passing focus to the explorer. If I change that boolean to false, there will be no loop.
Currently explorer is designed in such a way that when focus is passed if there is no active editor that it will try to open the editor - we can potentially also change that.

Sorry for the slow response.

**Comment by bpasero** [comment]
@isidorn I pushed a change to not move focus into the explorer. I think that is a good change anyway because otherwise focus gets lost from the editor without explicit user choice.

**Comment by bpasero** [comment]
@LeNouveau can we try another thing with node.js, which is to `fs.open()` the path and then `fs.read()` it and see if an error is returned?

**Comment by isidorn** [comment]
@bpasero cool, thanks

**Comment by LeNouveau** [comment]
@bpasero 

The script : 
```js
fs = require('fs')

fs.open('Simple gsheet file.gsheet', 'w', 0o666, (err, fd) => {
    if (err) {
        console.error(err);
    } else {
        fs.fstat(fd, (err2, stats) => {
            if (err2) {
                console.error(err2);
            } else {
                var bufferSize = stats.size,
                    chunkSize = 512,
                    buffer = Buffer.alloc(bufferSize),
                    bytesRead = 0;

                while (bytesRead < bufferSize) {
                    if ((bytesRead + chunkSize) > bufferSize) {
                        chunkSize = (bufferSize - bytesRead);
                    }
                    fs.read(fd, buffer, bytesRead, chunkSize, bytesRead);
                    bytesRead += chunkSize;
                }
                console.log(buffer.toString('utf8', 0, bufferSize));
                fs.close(fd, (err3) => {
                    if (err3) {
                        console.error(err3)
                    }
                });
            }
        });
    }
})

```

The result : 
```
PS G:\Mon Drive\test> node .\bla.js
OK read

Ready to close
```

As it is a `*.gsheet` file on the sync Drive, nothing is saved inside, even if I fill lines. 
It's always 1ko. 
It's like a link, it opens the browser.

**Comment by bpasero** [comment]
@LeNouveau any difference if you open with `r` flag instead of `w`?

**Comment by LeNouveau** [comment]
@bpasero No, exact same output, with `r` flag. 

Google Drive File Stream version : 29.1.85.2056
(for information, if results evolve)

**Comment by LeNouveau** [comment]
Sorry, miss click

**Comment by bpasero** [comment]
ok super weird, I am out of ideas why we think the files are folders :-/

**Comment by lcharbon** [comment]
Can this be reconsidered considering the additional duplicate issues?
